test/pmap_test.dart

```diff
@@ -3,24 +3,39 @@ import 'package:test/test.dart';
 
 int _square(int x) => x * x;
 
-Iterable<int> _countTo(int x) sync* {
-  for( int i = 0; i < x; ++i) {
-    yield i;
-  }
-}
+int? _maybeSquare(int? x) => x != null ? x * x : null;
 
 void main() {
-  test('square parallel 1', () async {
-    List<int> input = _countTo(1000).toList();
-    List<int> mapped = input.map(_square).toList();
-    List<int> pmapped = await pmap(input, _square).toList();
-    expect(pmapped, equals(mapped));
-  });
+  final input = Iterable<int>.generate(1000);
```

Comment:
Can you add a test that hits the `allElementsPublished = false` case?  Maybe throw an exception if a certain number input is given.

Reply:
That was quite the oversight.

Unfortunately the result of trying to add it is a little sad:

* Throwing for some values just returns the error to the main thread (and halts the calculation, but does not exit)
* `exit`ing will kill the whole process not only the Isolate
* killing the Isolate results in the same issue as throwing – meaning the calculation halts but does not exit

If you want to try for yourself:

```dart
import 'dart:io';
import 'dart:isolate';

import 'package:pmap/pmap.dart';

int _squareWithIssues(int x) => x == 53
    ? () {
        throw Exception('test');
        // exit(1);
        // Isolate.current.kill();
        return -1;
      }()
    : x * x;

void main() async {
  final foo = Iterable<int>.generate(100000);
  foo.mapParallel(_squareWithIssues, parallel: 2).listen(print);
}
```
**My suggestion would be to remove the test logic from `pmap` for now, merge and release the update, and improve error handling afterwards. Do you agree?**

* The current PR is already really big, 
* the old version did not handle that case either
* the old version lacks null-safety
* the old version has the issue with ordering, which is quite toxic (as it silently drops values without errors)

(If we add proper error handling I personally feel like a `1.0.0` release is earned :D)